state-chain/pallets/cf-elections/src/lib.rs

```diff
@@ -0,0 +1,1955 @@
+		StorageValue<_, ElectoralSystemStatus, OptionQuery>;
+
+	// ---------------------------------------------------------------------------------------- //
+
+	mod access_impls {
+		use super::*;
+
+		/// Implements traits to allow electoral systems to read/write an Election's details.
+		pub struct ElectionAccess<T: Config<I>, I: 'static> {
+			election_identifier: ElectionIdentifierOf<T::ElectoralSystem>,
+			_phantom: core::marker::PhantomData<(T, I)>,
+		}
+		impl<T: Config<I>, I: 'static> ElectionAccess<T, I> {
+			pub(super) fn new(
+				election_identifier: ElectionIdentifierOf<T::ElectoralSystem>,
+			) -> Self {
+				Self { election_identifier, _phantom: Default::default() }
+			}
+
+			fn unique_monotonic_identifier(&self) -> UniqueMonotonicIdentifier {
+				*self.election_identifier.unique_monotonic()
+			}
+		}
+		impl<T: Config<I>, I: 'static> ElectionReadAccess for ElectionAccess<T, I> {
+			type ElectoralSystem = T::ElectoralSystem;
+
+			fn settings(
+				&self,
+			) -> Result<
+				<T::ElectoralSystem as ElectoralSystem>::ElectoralSettings,
+				CorruptStorageError,
+			> {
+				let mut settings_boundaries =
+					ElectoralSettings::<T, I>::iter_keys().collect::<Vec<_>>();
+				settings_boundaries.sort();
+				let settings_boundary = settings_boundaries
+					.iter()
+					.rev()
+					.find(|settings_boundary| {
+						**settings_boundary <= self.unique_monotonic_identifier()
+					})
+					.ok_or(CorruptStorageError)?;
+				ElectoralSettings::<T, I>::get(settings_boundary).ok_or(CorruptStorageError)
+			}
+			fn properties(
+				&self,
+			) -> Result<
+				<T::ElectoralSystem as ElectoralSystem>::ElectionProperties,
+				CorruptStorageError,
+			> {
+				ElectionProperties::<T, I>::get(self.election_identifier).ok_or(CorruptStorageError)
+			}
+			fn state(
+				&self,
+			) -> Result<<T::ElectoralSystem as ElectoralSystem>::ElectionState, CorruptStorageError>
+			{
+				ElectionState::<T, I>::get(self.unique_monotonic_identifier())
+					.ok_or(CorruptStorageError)
+			}
+		}
+		impl<T: Config<I>, I: 'static> ElectionWriteAccess for ElectionAccess<T, I> {
+			fn set_state(
+				&mut self,
+				state: <T::ElectoralSystem as ElectoralSystem>::ElectionState,
+			) -> Result<(), CorruptStorageError> {
+				if self.state()? != state {
+					ElectionState::<T, I>::insert(self.unique_monotonic_identifier(), state);
+					ElectionConsensusHistoryUpToDate::<T, I>::remove(
+						self.unique_monotonic_identifier(),
+					);
+				}
+
+				Ok(())
+			}
+			fn clear_votes(&mut self) {
+				let unique_monotonic_identifier = self.unique_monotonic_identifier();
+				ElectionBitmapComponents::<T, I>::clear(unique_monotonic_identifier);
+				for (_, (_, individual_component)) in
+					IndividualComponents::<T, I>::drain_prefix(unique_monotonic_identifier)
+				{
+					<<T::ElectoralSystem as ElectoralSystem>::Vote as VoteStorage>::visit_shared_data_references_in_individual_component(&individual_component, |shared_data_hash| {
```

Comment:
Pedantic point - As far as I can tell, the `visit_shared_data_references` methods actually visit the hash, even thought they are always expected to either increment or decrement the reference count.

For example, it could be used to access the shared data itself (probably not very useful). My point is that maybe the closure should take a ReferenceDetails struct. Or even simpler, we add methods add/remove a reference.